src/core/mount.c

```diff
@@ -2079,25 +2076,17 @@ static int mount_process_proc_self_mountinfo(Manager *m) {
 
                 if (mount_is_mounted(mount) &&
                     mount->from_proc_self_mountinfo &&
-                    mount->parameters_proc_self_mountinfo.what) {
-                        /* Track devices currently used */
-
-                        if (set_ensure_allocated(&around, &path_hash_ops) < 0 ||
-                            set_put_strdup(&around, mount->parameters_proc_self_mountinfo.what) < 0)
-                                log_oom();
-                }
+                    mount->parameters_proc_self_mountinfo.what)
+                        /* Exclude devices currently used */
+                        free(set_remove(gone, mount->parameters_proc_self_mountinfo.what));
 
                 /* Reset the flags for later calls */
                 mount->proc_flags = 0;
         }
 
-        SET_FOREACH(what, gone) {
-                if (set_contains(around, what))
-                        continue;
```

Comment:
this doesn't look right.

If a specific device is mounted to multiple places, and one of them is removed, then we should not mistake this for it being gone entirely...

That's why the old code maintained two lists: of the devices definitely around, and the ones that appear gone. But if one of the gone devices actually appears in the around list, we'd know it's not actually gone...